File: frontend/app/types/movieReview.ts
import type { MovieSummary } from './movie'

export interface ReviewAuthor {
  id: string
  username: string
  display_name: string
  profile_picture: string | null
}

export interface MovieReview {
  id: string
  user_id: string
  movie_id: number
  rating: number
  review: string
  movie: MovieSummary
  author: ReviewAuthor
  created_at: string
  updated_at: string
}

export interface MovieReviewInput {
  rating: number
  review: string
}

export interface MovieReviewListResponse {
  results: MovieReview[]
}

File: backend/app/schemas/movie_review.py
"""Validation and response contracts for user-authored movie reviews."""

from datetime import datetime

from pydantic import BaseModel, Field, field_validator

from app.schemas.movie import MovieSummaryResponse


class MovieReviewUpsertRequest(BaseModel):
    rating: float = Field(ge=1, le=10)
    review: str = Field(min_length=1, max_length=5_000)

    @field_validator("review")
    @classmethod
    def review_must_contain_text(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("Review must contain text")
        return normalized


class ReviewAuthorResponse(BaseModel):
    id: str
    username: str
    display_name: str
    profile_picture: str | None = None


class MovieReviewResponse(BaseModel):
    id: str
    user_id: str
    movie_id: int
    rating: float
    review: str
    movie: MovieSummaryResponse
    author: ReviewAuthorResponse
    created_at: datetime
    updated_at: datetime


class MovieReviewListResponse(BaseModel):
    results: list[MovieReviewResponse]
